### ISSUE TITLE
Refactor model class hierarchy

### ISSUE THREAD

**Issue by joeloskarsson** [mllam/neural-lam#49]
## Background
The different models that can be trained in Neural-LAM are currently all sub-classes of `pytorch_lightning.LightningModule`. In particular, much of the functionality  sits in the first subclass, `ARModel`. The current hierarchy looks like this:
![classes_current](https://github.com/mllam/neural-lam/assets/16475490/0cfdead6-2971-4a3b-b42f-8d80fe850c0d)
(I am making these rather than some more fancy UML-diagrams since I think this should be enough for the level of detail we need to discuss here).

## The problem
In the current hierarchy everything is a subclass of `ARModel`. This has a number of drawbacks:

* All functionality ends up in one class. There is no clear division of responsibility, but rather we end up with one class that does everything from logging, device-handling, unrolling forecasts and the actual forward and backward pass through GNNs. 
* The subclasses do not utilize the standard torch `forward` calls, but rather must resorts to our own similar construction (e.g. `predict_step`)
* This limits us to deterministic, auto-regressive models.
* This is hard to extend upon for ensemble-models.

## Proposed new hierarchy
I propose to split up the current class hierarchy into subclasses that have clear responsibilities. These should not just all inherit `ARModel`, but rather be members of each other as suitable. A first idea for this is shown below, including also potential future classes for new models (to show how this is more extendible):

![classes_proposal](https://github.com/mllam/neural-lam/assets/16475490/53bc4e7d-abf0-44b8-9819-ed59d4e3ca89)

The core components are (I here count static features as part of forcing):

* `ForecasterModule`: Takes over much of the responsibility of the old `ARModel`. Handles things not directly related to the nerual network components such as plotting, logging, moving batches to the right device. This inherits `pytorch_lightning.LightningModule` and have the different train/val/test steps. In each step (train/val/test), unpacks the batch of tensors and uses a `Forecaster` to produce a full forecast. Also responsible for computing the loss based in a produced forecast (could also be in `Forecaster`, not entirely sure about this).
* `Forecaster`: A generic forecaster capable of mapping from a set of initial states, forcing and boundary forcing into a full forecast of the requested length.
  * `ARForecaster`: Subclass of `Forecaster` that uses an auto-regressive strategy to unroll a forecast. Makes use of a `StepPredictor` at each AR step.
* `StepPredictor`: A model mapping from the two previous time steps + forcing + boundary forcing to a prediction of the next state. Corresponds to the $\hat{f}$ function in [Oskarsson et al.](https://arxiv.org/abs/2309.17370).
  * We find the existing graph models now as subclasses to `StepPredictor`.

In the figure above we can also see how new kinds of models could fit into this hierarchy:

* Ensemble models using an auto-regressive strategy to sample the state at each time step.
* Other types of single-step-predictors, e.g. using CNNs or Vision-Transformers.
* Non-autoregressive strategys for creating a full forecast, in particular direct forecasting.

This is supposed to be a starting point for discussion and there will likely be things I have not thought about. Some parts of this will have to be hammered out when actually writing these classes, but I'd rather have the discussion whether this is a good direction to take things before starting to do too much work. Tagging @leifdenby and @sadamov for visibility.

**Comment by sadamov** [comment]
@joeloskarsson This looks great! Introducing even more freedom and clearly structured and understandable classes. I am no expert in this so not adding too much here. I really think though, that relieving `ar_models.py` from some of its duties will make our lives easier. Also, you already have experience with probabilistic Neural-LAM so I trust you to make the changes necessary to implement that in the future. :heart:

**Comment by leifdenby** [comment]
I have some time to give this a proper think through now @joeloskarsson so here come my thoughts :)

> * All functionality ends up in one class. There is no clear division of responsibility, but rather we end up with one class that does everything from logging, device-handling, unrolling forecasts and the actual forward and backward pass through GNNs.
> 
> * The subclasses do not utilize the standard torch `forward` calls, but rather must resorts to our own similar construction (e.g. `predict_step`)
> 
> * This limits us to deterministic, auto-regressive models.
> 
> * This is hard to extend upon for ensemble-models.

Yes, agreed! Thank you for making that diagram too. It is very helpful in understanding the current structure.

> * `ForecasterModule`: Takes over much of the responsibility of the old `ARModel`. Handles things not directly related to the nerual network components such as plotting, logging, moving batches to the right device. This inherits `pytorch_lightning.LightningModule` and have the different train/val/test steps. In each step (train/val/test), unpacks the batch of tensors and uses a `Forecaster` to produce a full forecast. Also responsible for computing the loss based in a produced forecast (could also be in `Forecaster`, not entirely sure about this).

This sounds good. My only suggestion would be to not include the plotting as a method of the `ForecasterModule` or `Forecaster` classes, but instead keep this a separate (and therefore stateless) function. Keeping plotting separately would allow us to make the model architecture classes more lean and design for reuse of plotting functions. Regarding where to store the loss operation: I think I would place this with `Forecaster` since that class only handles forecasting for a given time-window, whereas `ForecasterModule` would also handle the data-handling used for constructing time-windows.

I think the structure you have proposed with `ARForecaster`, `StepPredictor` and the current (and future) graph-based models makes a lot of sense too, I really like it. And I see how new architectures can be encorporated, but not completely. I was thinking for example for the `BaseGraphModel` family of methods vs a future `CNNPredictor`, these will have different requirements in terms of what a single batch should contain (spatial coordinates stacked or not). Have you thought about at which level this will impact the `forward` methods of this new hierarchy? Maybe this is self-evident, but to me it is quite tricky. It might simply be a question of well-written docstrings and a convention (in terms of which dimensions in the input tensor represent what) that we apply through the code. Minor thought: should it maybe be `BaseGraphModelPredictor` rather than `BaseGraphModel` or something like? Just to follow the `*Predictor` name convention you are setting up for here.

Maybe when you/we start coding this we could start by including the additions you forse making (ensemble prediction classes, CNN, ViT) as simply empty placeholder classes? Then we can be sure the data-structures we pass around will be general enough to apply in future too?

In general this is really fantastic though.

**Comment by joeloskarsson** [comment]
Thanks for some very valuable input @leifdenby!

> My only suggestion would be to not include the plotting as a method of the `ForecasterModule` or `Forecaster` classes, but instead keep this a separate (and therefore stateless) function.

Yes, that is what I have in mind as well. In the same way it is handled today, where `ARModel` just makes calls to plotting functions in `neural_lam.vis`. Potentially even more surrounding logic could be moved away from the `ForecasterModule`, but will see if this is would be the correct PR for that.

> Regarding where to store the loss operation: I think I would place this with `Forecaster` since that class only handles forecasting for a given time-window, whereas `ForecasterModule` would also handle the data-handling used for constructing time-windows.

There is a logic to that yes. I am still thinking about how to do this in practice. I am in a way not particularly happy to send the target (ground truth) tensor further down the hierarchy than `ForecasterModule`, which would be required to compute the loss in `Forecaster`. In practice one would either have to 1) Let the `forward` function of `Forecaster` map from inputs + target tensor to the loss value. This violates what one would expect a `forward` call to do (produce the output/prediction). 2) Have a separate function in `Forecaster` for computing the loss, which means we have to again define our own interface for parts of this, rather than relying completely on Lightnings `train_step`. It's a bit tricky, maybe easiest to start hacking away at and see what becomes the most convenient.

> I was thinking for example for the `BaseGraphModel` family of methods vs a future `CNNPredictor`, these will have different requirements in terms of what a single batch should contain (spatial coordinates stacked or not). Have you thought about at which level this will impact the `forward` methods of this new hierarchy? Maybe this is self-evident, but to me it is quite tricky. It might simply be a question of well-written docstrings and a convention (in terms of which dimensions in the input tensor represent what) that we apply through the code. 

Yes, this is a good point. I have given this some thought. One should note that having 1 (flattened) spatial dimensions or 2 (x,y) should not majorly impact anything before the `StepPredictors`. Since moving between these is just a cheap reshape, it is mostly a matter of convention. You can either flatten already in the dataset class and then reshape to 2 spatial dims. in e.g. CNN-models (current setup), or you keep your spatial dims. from the dataloader and flatten in the graph-based models. This is something that should not be very hard to change later on if needed.

Here is an example of a dummy CNN-class I have laying around that handles this with reshapes (using old class hierarchy):
```python
import torch

from neural_lam.models.ar_model import ARModel
from neural_lam import constants

class NewModel(ARModel):
    """
    A new auto-regressive weather forecasting model
    """
    def __init__(self, args):
        super().__init__(args)

        # Some dimensionalities that can be useful to have stored
        self.input_dim = 2*constants.grid_state_dim + constants.grid_forcing_dim +\
            constants.batch_static_feature_dim
        self.output_dim = constants.grid_state_dim

        # TODO: Define modules as members here that will be used in predict_step
        self.layer = torch.nn.Conv2d(self.input_dim, self.output_dim, 1) # Dummy layer

    def predict_step(self, prev_state, prev_prev_state, batch_static_features, forcing):
        """
        Predict weather state one time step ahead
        X_{t-1}, X_t -> X_t+1

        prev_state: (B, N_grid, d_state), weather state X_t at time t
        prev_prev_state: (B, N_grid, d_state), weather state X_{t-1} at time t-1
        batch_static_features: (B, N_grid, batch_static_feature_dim), static forcing
        forcing: (B, N_grid, forcing_dim), dynamic forcing

        Returns:
        next_state: (B, N_grid, d_state), predicted weather state X_{t+1} at time t+1
        pred_std: None or (B, N_grid, d_state), predicted standard-deviations
                    (pred_std can be ignored by just returning None)
        """

        # Reshape 1d grid to 2d image
        input_flat = torch.cat((prev_state, prev_prev_state, batch_static_features,
            forcing), dim=-1) # (B, N_grid, d_input)
        input_grid = torch.reshape(input_flat, (-1, *constants.grid_shape,
            input_flat.shape[2])) # (B, N_x, N_y, d_input)
        # Most computer vision methods in torch want channel dimension first
        input_grid = input_grid.permute((0,3,1,2)).contiguous() # (B, d_input, N_x, N_y)

        # TODO: Feed input_grid through some model to predict output_grid
        output_grid = self.layer(input_grid) # Shape (B, d_state, N_x, N_y)

        # Reshape back from 2d to flattened grid dimension
        output_grid = output_grid.permute((0,2,3,1)) # (B, N_x, N_y, d_state)
        next_state = output_grid.flatten(1,2) # (B, N_grid, d_state)

        return next_state, None
```

A good reason to keep it as is, flattening in the Dataset class, is that once we start moving to more refined boundary setups, on different gridding, there will no longer be a 2d grid representation of the input data. Grid cells will be on irregular grids, meaning that it is not trivial to apply CNN models to them. Keeping things flattened all the way to the CNN model means that if you want to define a CNN on this you have to decide how to handle this irregularity. We don't have to make such decisions before the `forward` call in the CNN class.

> Minor thought: should it maybe be `BaseGraphModelPredictor` rather than `BaseGraphModel` or something like? Just to follow the `*Predictor` name convention you are setting up for here.

Yes, that is a good idea. Or even `BaseGraphPredictor` and `BaseHiGraphPredictor`. Not sure if one would want to also name the final models (e.g. HiLAM) something like `HiLAMPredictor`, just to specify that these are also `StepPredictor`s? On the other hand, it's maybe best to keep to the exact naming of a model used in e.g. a paper.

> Maybe when you/we start coding this we could start by including the additions you forse making (ensemble prediction classes, CNN, ViT) as simply empty placeholder classes? Then we can be sure the data-structures we pass around will be general enough to apply in future too?

This could be smart yes. For the computer-visions models (CNN, ViT) I don't think that there will be much to keep in mind (see dicsussion above), but for the ensemble model this could make a lot of sense. Especially since I anyhow will populate that with the code from the probabilistic model branches later. My plan is to do this refactoring first, to make the merging of that easier and nicer.

Overall it looks like there is support for this idea, so I can start writing the code for it. Then we can discuss more details in upcoming PR. I am happy to do the work on this, but I have a hard time to give a timeline, as it is not directly crucial to progress in ongoing research projects. Anyhow, I think this could potentially fit in v0.3.0 in the roadmap?

**Comment by leifdenby** [comment]
I have finally started work on this :) I really want to try out @ErikLarssonDev's diffusion models on Gefion before we loose access at the end of the month.

I am working on the following branch (although don't look yet, this is very much work-in-progress, I am still setting up the module/class structures): https://github.com/leifdenby/neural-lam/tree/feat/model-archs-refactor

**Comment by leifdenby** [comment]
Refactoring is progressing 😄 I wrote [some code](https://github.com/leifdenby/neural-lam/blob/feat/model-archs-refactor/docs/class_diagram_gen.py) (with the help of chatgpt) to generate a class-diagram directly from the classes I have implemented:

![Image](https://github.com/user-attachments/assets/7bf7e5c4-960c-473e-b0ad-c099105e92e7)

I was thinking this might be neat to include into the documentation once we build that :)

**Comment by Erik-Wikingsson** [comment]
@leifdenby 

I've implemented a ARProbModel class in https://github.com/ErikLarssonDev/real-prob-lam and the Diffusion-LAM model. The Graph-EFM model implementation is in progress.

If you want to look at the code, let me know, and I can give you access.

This work builds upon https://github.com/joeloskarsson/neural-lam-dev/releases/tag/building-ml-lams by adding probabilistic modelling.

This is related to https://github.com/mllam/neural-lam/issues/62 but I will write a similar update here as well. 

I have done:
- Implemented cropping for a rectangular mesh grid (enabling processing with a non-graph-based processor) https://github.com/ErikLarssonDev/mllam-data-prep-real/tree/rect-boundary
- A graph/U-Net model that encodes the grid to a rectangular mesh with a GNN, processes the mesh with a U-Net processor, and then decodes the mesh to the grid with a GNN. This could be extended to any processor architecture that works on a regular grid.
- Implemented Diffusion-LAM with the Graph-U-Net architecture, but it could also be run with a Graph-FM model if the user wants a purely graph-based model.
- For now, I am working on the research project and finishing the implementation of Graph-EFM, but when I am done, I think it would be nice to merge it into the main branch. If you want to start taking part now, that is fine as well.

**Comment by Erik-Wikingsson** [comment]
@leifdenby I also have something similar for the MEPS data, along with a few other probabilistic models, here: https://github.com/ErikLarssonDev/neural-lam

That said, I think it’s easier to work from the realistic-setting branch, since this repository is a bit more of a “wild west” playground where I write some quick and dirty code to test things quickly. 🙂

I will clean this up a little bit eventually as I am working on it, but I will not bring it to the main branch since this is code for the MEPS dataset, and I want to keep it as that. If I want to bring any of these models or code developments to a larger setting, I will write some better code in the realistic probabilistic repository instead.

**Comment by AdMub** [comment]
Hi @joeloskarsson and @leifdenby,

I’ve been spending the last few days deep in the `weather-model-graphs` architecture (fixing some of the spatial pruning, `g2m` safety checks, and `cugraph` backend bugs over there) and I am very interested in helping drive this `neural-lam` refactor forward for GSoC.

Looking at the proposed `ForecasterModule` -> `Forecaster` -> `StepPredictor` hierarchy, I agree this is the exact right move to support the generative/ensemble architectures coming in #62.

@leifdenby - I saw you started laying the groundwork for this in the `feat/model-archs-refactor` branch. Because I’m coming fresh off the graph generation side, I’ve been thinking specifically about Joel's point regarding how spatial dimensions (flattened 1D graphs vs. 2D grids for CNNs) should impact the `forward` methods of the new `StepPredictor` subclasses.

If we keep the data flattened in the dataloader (to support irregular meshes seamlessly), the `BaseGraphPredictor` will take it natively, and any future `CNNPredictor` can handle the `(B, N_x, N_y, d_input)` reshape internally as Joel prototyped.

**How can I best assist you with the `feat/model-archs-refactor` branch right now?** 
1. Would it be helpful if I checked out your branch and started stubbing out the `BaseGraphPredictor` and `BaseHiGraphPredictor` classes so they correctly interface with the `weather-model-graphs` outputs?
2. Or would you prefer I start extracting the plotting functions from the old `ARModel` into stateless functions in `neural_lam.vis` as discussed above?

Happy to jump into whichever piece unblocks this hierarchy refactor the fastest!

**Comment by aniket2405** [comment]
@joeloskarsson 
Wanted to check if we want to put this for work specifically during GSOC or are you open to contributions before that?
Additionally, we can split this into multiple PRs for better readability.
Happy to talk further on this.

**Comment by joeloskarsson** [comment]
Going through with a larger refactoring like this requires quite a lot of design decisions, and would be most suitable as part of GSoC work within project 3. But it is also true that this would best be split into multiple PRs. So if there is interest one could do some smaller pieces of this work now, to also get acquainted with the code. 

Extracting parts of the evaluation and plotting code from the existing `ARModel` into stateless functions could indeed be a good starting point for this.

**Comment by AdMub** [comment]
Thanks for the clarity, @joeloskarsson! That makes perfect sense. Given the scope of the design decisions required for the `ForecasterModule` -> `Forecaster` -> `StepPredictor` hierarchy, I agree this is definitely the core work for GSoC Project 3 rather than a quick weekend PR.

Instead of duplicating the plotting extraction work that others have already started below, I am going to focus my energy on mapping out the exact, multi-PR rollout strategy for this class hierarchy refactor in my GSoC Proposal Document. I'll make sure the proposal specifically addresses how the forward calls will handle the flattened spatial dimensions to accommodate both graph and future CNN/ViT architectures seamlessly.

Looking forward to sharing the draft with you and @leifdenby soon!

**Comment by mandeepsingh2007** [comment]
https://github.com/mllam/neural-lam/pull/246
@joeloskarsson @leifdenby i have just opened a PR that extracts the plotting and evaluation components from the ARModel class into stateless functions in a new module rhis breaks down the monolithic ARModel and improves modularity and testability.

now what should i focus on next for gsoc project 3

**Comment by techaadii** [comment]
Hey I would love to work on this problem statement

**Comment by joeloskarsson** [comment]
Some work on this refactoring, for deterministic models, has started in #208 . For those who want to contribute, any comments and ideas are welcome there! 😄

**Comment by Debadri-das** [comment]
HI @joeloskarsson , would love to contribute to this problem statement.

**Comment by AdMub** [comment]
Hi @Sir-Sloth-The-Lazy, great work getting the foundation of the `ForecasterModule` -> `ARForecaster `-> `StepPredictor` hierarchy laid out! 
Looking at this from the perspective of the upcoming probabilistic integration (Project 3 / #62), I have one architectural thought:
How rigid is the return signature of the `StepPredictor.forward()` currently? To ensure this seamlessly supports @ErikLarssonDev's `ARProbModel` and diffusion models later, we should ensure the `StepPredictor` interface allows for returning parameter distributions (e.g., mean and variance tensors) rather than just deterministic state predictions. If we enforce a strict (`next_state`) return now, we will have to rewrite the interface in Phase 2.

I am mapping out the exact integration phases for these probabilistic models in my GSoC proposal, and having this deterministic baseline solidifies Phase 1 perfectly.

**Comment by AdityaKumarSethia** [comment]
## Proposal: EnsembleForecaster — extending this hierarchy for probabilistic forecasting (GSoC Project 3)
Hi @joeloskarsson, @leifdenby  — thanks for the detailed discussion here. I've been following both this issue and PR #208 closely, as I'm targeting GSoC Project 3 (Generalizing to probabilistic forecasting models). I'd love to share a concrete design idea and get your early feedback on whether it fits the direction you're envisioning.
### The idea: EnsembleForecaster(BaseForecaster)
The composable `Forecaster` / `StepPredictor` split introduced in this PR creates a natural extension point for ensemble and probabilistic models. I'm proposing an `EnsembleForecaster` that inherits from `Forecaster` and holds a single `StepPredictor` instance, running it `N` times with different noise samples to produce a stochastic ensemble:
```python
class EnsembleForecaster(BaseForecaster):
    """
    Produces n_members ensemble members from a single StepPredictor by
    injecting independent noise at each rollout. The predictor is encoded
    once per timestep; diversity comes purely from the sampled noise ε_i ~ N(0, I).

    Args:
        predictor (BaseStepPredictor): The underlying single-step model.
        n_members (int): Number of ensemble members to sample.
    """
    def __init__(self, predictor: BaseStepPredictor, n_members: int):
        super().__init__(predictor)
        self.n_members = n_members

    def forward(
        self,
        init_states: torch.Tensor,       # B, S_init, grid, d_state)
        forcing_features: torch.Tensor,  # (B, T, grid, d_forcing)
        boundary_states: torch.Tensor,   # (B, T, grid, d_state)
    ) -> torch.Tensor:
        # Returns (B, S, T, N, F) — S ensemble members, N spatial nodes, F features
        members = []
        for _ in range(self.n_members):
            members.append(
                self._ar_rollout(init_states, forcing_features, boundary_states)
            )
        return torch.stack(members, dim=1)
```
The diagram below shows where EnsembleForecaster fits in the proposed hierarchy (highlighted in purple) :

<img width="590" height="662" alt="Image" src="https://github.com/user-attachments/assets/d60607f0-4a09-40a4-84d5-65ae93e8a7b2" />

I also came across #335 which proposes standardizing the forecast tensor shape contract. `EnsembleForecaster` would return `(B, S, T, N, F)` as defined there, and could serve as the first concrete implementation that validates whether that contract is workable end-to-end — covering both the rollout side (PR #208) and the evaluation/plotting side (#335).

One design question I'd love your input on: should `EnsembleForecaster` reuse a shared `_ar_rollout` from `BaseForecaster`, or should ensemble unrolling be implemented independently to keep the single-trajectory path clean?

**Comment by kshirajahere** [comment]
@joeloskarsson one thing I am still trying to pin down in #49 is where you would want the first ensemble-specific extension point to live once the deterministic split from #208 settles.

Would you prefer probabilistic support to enter as a dedicated `ForecasterModule` subclass that changes loss / metric routing, while keeping `ARForecaster` and `StepPredictor` single-trajectory? Or do you see a dedicated `Forecaster` subclass as the cleaner long-term boundary?

I am asking because the two choices seem to imply quite different responsibilities for the core forecasting path, and I want to make sure I am reading the intended direction correctly before baking that into Project 3 planning.

**Comment by XVX-016** [comment]
Hi, I'm Tanmmay, a GSoC 2026 applicant who has submitted a proposal for the "Generalizing to Probabilistic Forecasting Models" project. I've been reading through this issue and the linked papers (arXiv:2406.04759 and arXiv:2502.07532) to understand the existing prototype approaches.
A few questions to make sure my proposal aligns with the maintainers' vision:
- Between the score-based diffusion approach and the flow-matching approach, is there a preferred direction for the GSoC project, or should both be supported by the refactored interface?
- For the sample_forecast() interface, is there a preference for where this lives in the base ARModel class, or as a mixin?

I've also identified issue #62 as closely related. Happy to discuss the design further here or on Slack once I get an invite.

**Comment by Debadri-das** [comment]
Hi @joeloskarsson @leifdenby ,

The anticipated inclusion of probabilistic models from Issue #62 and this refactoring effort have been actively monitored by me. I saw two potential bottlenecks/edge situations that we may solve during this redesign when examining the EnsembleForecaster proposal in this thread and the way sample_trajectories is now handled in the prob_model_lam branch:

**1. O(S) Sequential Unrolling Bottleneck**
A sequential PyTorch for loop over self.n_members is currently used for the ensemble formation (for example, for _ in range(self.ensemble_size): unroll()). This O(S) sequential unrolling will significantly underuse the GPU and slow down inference for higher standard ensemble sizes (e.g., S=50).

 **-Proposed Solution:** If the StepPredictor is pure PyTorch, we can perform batched unrolling. We can expand the initial states from (B, ...) to (B * S, ...) out-of-the-box using torch.repeat_interleave() before feeding it into the AR loop. After the unroll is entirely completed in a single batched pass, we reshape it back to (B, S, pred_steps, ...). If memory (OOM) is a concern for B×S, we can implement a chunk_size argument to unroll in minibatches.


**2. Ensemble Lateral Boundary Conditions (LBCs)**
In the current unroll_prediction logic, the boundary points are overwritten by true_states (the deterministic forcing). If done across an ensemble, every member S receives the exact same deterministic boundary condition, which in meteorological contexts artificially crashes the ensemble spread to zero near the domain edges.

**-Proposed Solution:** In the new ForecasterModule / EnsembleForecaster interface, we should allow boundary_states to accept an optional ensemble dimension: (B, S, time, grid, d_state). If a dataset provides probabilistically perturbed boundaries (which is standard for operational LAMs), the model can index border_state[:, s, i] rather than clamping all members to a single sequence.


When the probabilistic models are combined, I believe that setting up the API for these two items now will save a lot of difficulties. If you agree this is the best course of action, I'd be pleased to create a brief draft design or start a tiny PR with an EnsembleForecaster prototype showing the batched unrolling!

**Comment by Anish0104** [comment]
I'm applying for GSoC 2026 Project 3 and have been studying this issue. My plan is to generalize the model interface by introducing a sample() method alongside forward() in the base model class - deterministic models use only forward() and remain unchanged, while probabilistic models implement both. This cleanly separates training (loss computation) from inference (ensemble sampling). Happy to discuss the design before the proposal deadline.

**Comment by Debadri-das** [comment]
Autoregressive (AR) ensemble forecasting is carried out by ar_rollout_batched(), which processes all ensemble members in a single vectorized loop rather than in many iterations. By extending the batch dimension from B → B*S, the core optimization enables the model to forecast all S ensemble members simultaneously in a single time loop.

Link: https://gist.github.com/Debadri-das/9b887e41b7f29e0b759ea432f0df372a

### Problem It Solves:

This produces B*S*T separate forward passes, which is inefficient on GPUs that excel at parallel batch processing. ar_rollout_batched() collapses this to a single O(T) loop where each iteration processes all B*S samples simultaneously.

Traditional AR ensemble unrolling requires nested loops:
```python
      for t in range(T):
        for s in range(S):
            pred = model(state, forcing)  # Called B*S*T times

**Comment by SUMIT-SHARMA-04** [comment]
Hi team, Sumit Sharma here! I’ve been looking for a solid starting point to contribute to MLLAM, and I'd love to take a crack at refactoring the model class hierarchy.

I have a solid background in machine learning and I'm currently familiarizing myself with the codebase. Before I open a draft PR, could anyone point me to the specific files that need the most attention or share any preferred design patterns for the new hierarchy? Happy to get started as soon as I have a bit more context!

**Comment by AST119** [comment]
Hi everyone 👋
I'm Aaditya Singh and my domain is machine learning and python. I'm contributing 1st time as a GSOC Contributor and would like your help to make it easier to understand the right way to start contributing

I’ve been going through the proposed refactor and the ongoing discussions in detail, and I really like the direction of splitting responsibilities into **ForecasterModule → Forecaster → StepPredictor**. It makes the architecture much more modular and future-proof for probabilistic and generative extensions.

From what I understand, the core structure is mostly agreed upon, and current work is focused on:

* stabilizing interfaces (especially `forward()` vs custom methods),
* handling input shape flexibility (graph vs grid),
* and enabling efficient probabilistic/ensemble forecasting.

### My thoughts / proposal

I’d like to contribute by focusing on two areas that seem both impactful and currently open:

#### 1. Batched Ensemble Rollout (Performance Improvement)

Instead of sequentially looping over ensemble members, I propose implementing a batched rollout by expanding the batch dimension:

* `(B → B × S)` before AR rollout
* reshape back to `(B, S, T, N, F)` after

This would:

* fully utilize GPU parallelism
* align well with the proposed `EnsembleForecaster` design
* avoid O(S) slowdown for large ensembles

I can prototype this inside a reusable `_ar_rollout_batched()` method within `Forecaster`.

---

#### 2. Clean StepPredictor Interface (for CNN + Graph models)

To handle both flattened (graph) and grid (CNN/ViT) inputs cleanly, I suggest:

* keeping **flattened representation as default contract**
* letting each `StepPredictor` subclass internally reshape as needed

This keeps:

* dataloader simple
* core pipeline consistent
* flexibility for irregular meshes in future

---

#### 3. Small starter contributions (if helpful)

To get aligned with the current refactor branch, I can start by:

* stubbing `BaseGraphPredictor` and `BaseHiGraphPredictor`
* extracting plotting utilities into stateless functions
* adding docstrings for tensor shape conventions across modules

---
Please let me know which part would be most helpful to prioritize right now. I’d be happy to align with the ongoing `feat/model-archs-refactor` work and iterate from there.

Looking forward to contributing to this refactor

**Comment by aryannzzz** [comment]
For the EnsembleForecaster extension in the context of Project #3, I'm planning to handle the ensemble dimension in unroll_prediction by expanding B → B×S before the AR loop (as Debadri-das notes in #489) rather than sequential loops. 

So like should EnsembleForecaster inherit from ARForecaster directly, or sit as a sibling at the Forecaster level?
